File: src/kicad_xyce_plugin/simulation_parameters/hb_simulation_parameters.py
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

from ..netlist_parser import NetlistTopology
from .print_parameters import PrintParameters


@dataclass(frozen=True)
class HbSimulationParameters:

    frequencies: tuple[str, ...]
    harmonics: tuple[int, ...] = ()
    tahb: int | None = None
    selectharms: str | None = None
    startup_periods: int | None = None
    replace_ground: bool = True
    print_parameters: PrintParameters | None = None
    nonlin_options: dict[str, str] = field(default_factory=dict)
    linsol_options: dict[str, str] = field(default_factory=dict)

    @classmethod
    def from_xyce_directives(cls, directives: list[str]) -> "HbSimulationParameters" | None:
        # init defaults
        frequencies: list[str] = []
        harmonics: list[int] = []
        tahb = None
        selectharms = None
        startup_periods = None
        replace_ground = True
        print_parameters = None
        nonlin_options = {}
        linsol_options = {}
        # flag indicating whether a valid directive was found
        found = False
        # parse directives
        for directive in directives:
            # tokenize the directive
            tokens = directive.split()
            # skip empty directives
            if not tokens:
                # next
                continue
            # get command
            cmd = tokens[0].upper()
            # parse print directives and retain hb-specific output config
            if cmd == ".PRINT":
                # parse the print statement from the directive
                print_statement = PrintParameters.from_xyce_statement(directive)
                # retain hb print parameters when found
                if print_statement and print_statement.print_type in ("HB", "HB_FD", "HB_TD"):
                    # store the parsed print parameters
                    print_parameters = print_statement
                # next
                continue
            # handle preprocess replaceground
            if cmd == ".PREPROCESS" and len(tokens) > 2 and tokens[1].upper() == "REPLACEGROUND":
                # set replace_ground based on the third token
                replace_ground = tokens[2].upper() == "TRUE"
                # next
                continue
            # handle .OPTIONS
            if cmd == ".OPTIONS" and len(tokens) > 1:
                # get package
                pkg = tokens[1].upper()
                # handle HBINT
                if pkg == "HBINT":
                    # parse options
                    for token in tokens[2:]:
                        # check for equals
                        if "=" in token:
                            # split key and value
                            k, v = token.split("=", 1)
                            # normalize key
                            k = k.upper()
                            # handle NUMFREQ
                            if k == "NUMFREQ":
                                # split comma-separated values
                                try:
                                    # parse integers
                                    harmonics = [int(x) for x in v.split(",") if x.strip()]
                                # catch parse errors
                                except ValueError:
                                    # ignore invalid NUMFREQ
                                    pass
                            # handle TAHB
                            elif k == "TAHB":
                                # parse integer
                                try:
                                    # store tahb
                                    tahb = int(v)
                                # catch parse errors
                                except ValueError:
                                    # ignore
                                    pass
                            # handle SELECTHARMS
                            elif k == "SELECTHARMS":
                                # store selectharms
                                selectharms = v.lower()
                            # handle STARTUPPERIODS
                            elif k == "STARTUPPERIODS":
                                # parse integer
                                try:
                                    # store startup periods
                                    startup_periods = int(v)
                                # catch parse errors
                                except ValueError:
                                    # ignore
                                    pass
                    # next
                    continue
                # handle NONLIN-HB
                if pkg == "NONLIN-HB":
                    # parse options
                    for token in tokens[2:]:
                        # check for equals
                        if "=" in token:
                            # split key and value
                            k, v = token.split("=", 1)
                            # store
                            nonlin_options[k.upper()] = v
                    # next
                    continue
                # handle LINSOL-HB
                if pkg == "LINSOL-HB":
                    # parse options
                    for token in tokens[2:]:
                        # check for equals
                        if "=" in token:
                            # split key and value
                            k, v = token.split("=", 1)
                            # store
                            linsol_options[k.upper()] = v
                    # next
                    continue
            # skip non-HB directives
            if cmd != ".HB":
                # next
                continue
            # flag indicating a valid HB directive was found
            found = True
            # collect all fundamental frequencies from remaining tokens
            frequencies = tokens[1:]
        # return instance if a valid directive was found
        return cls(frequencies=tuple(frequencies), harmonics=tuple(harmonics), tahb=tahb, selectharms=selectharms, startup_periods=startup_periods, replace_ground=replace_ground, print_parameters=print_parameters, nonlin_options=nonlin_options, linsol_options=linsol_options) if found else None

    def to_xyce_directives(self, topology: NetlistTopology | None = None) -> list[str]:
        # prepend replaceground preprocessing when enabled
        preprocess = [".PREPROCESS REPLACEGROUND TRUE"] if self.replace_ground else []
        # build the hb directive with space-separated fundamental frequencies
        lines = [".HB " + " ".join(self.frequencies)]
        # build hbint options
        hbint_options: list[str] = []
        # append NUMFREQ when provided
        if self.harmonics:
            # format as comma-separated integers
            hbint_options.append("NUMFREQ=" + ",".join(str(h) for h in self.harmonics))
        # append TAHB when provided
        if self.tahb is not None:
            # append option
            hbint_options.append(f"TAHB={self.tahb}")
        # append SELECTHARMS when provided
        if self.selectharms:
            # append option
            hbint_options.append(f"SELECTHARMS={self.selectharms}")
        # append STARTUPPERIODS when provided
        if self.startup_periods is not None:
            # append option
            hbint_options.append(f"STARTUPPERIODS={self.startup_periods}")
        # check if we have hbint options
        if hbint_options:
            # append .OPTIONS HBINT line
            lines.append(".OPTIONS HBINT " + " ".join(hbint_options))
        # append nonlin-hb options
        if self.nonlin_options:
            # format as key=value pairs
            opts = " ".join(f"{k}={v}" for k, v in self.nonlin_options.items())
            # append directive
            lines.append(f".OPTIONS NONLIN-HB {opts}")
        # append linsol-hb options
        if self.linsol_options:
            # format as key=value pairs
            opts = " ".join(f"{k}={v}" for k, v in self.linsol_options.items())
            # append directive
            lines.append(f".OPTIONS LINSOL-HB {opts}")
        # append hb print directive when configured
        if self.print_parameters and self.print_parameters.print_type in ("HB", "HB_FD", "HB_TD"):
            # append the print statement
            lines.append(self.print_parameters.to_xyce_statement())
        # return the full directive list
        return preprocess + lines

    def raw_output_file_path(self, working_directory: Path, netlist_file_path: Path) -> Path | None:
        # check raw format is selected in print parameters and a print file is specified
        if self.print_parameters is None or self.print_parameters.print_format != "RAW":
            return None
        # return the output file path specified in the print parameters if available
        if self.print_parameters.print_file:
            return working_directory / self.print_parameters.print_file
        # otherwise, create raw file from netlist file, adding the ".raw" suffix
        return netlist_file_path.with_suffix(netlist_file_path.suffix + ".raw")

File: src/simulation_parameters/hb_simulation_parameters.cpp
#include <cctype>
#include <optional>
#include <string>
#include <vector>

#include "../util.h"
#include "hb_simulation_parameters.h"

HbSimulationParameters::HbSimulationParameters(std::vector<std::string> frequencies, std::vector<int> harmonics, std::optional<int> tahb, std::optional<std::string> selectharms, std::optional<int> startup_periods, std::optional<PrintParameters> print_parameters, std::map<std::string, std::string> nonlin_options, std::map<std::string, std::string> linsol_options) :
    frequencies(std::move(frequencies)), harmonics(std::move(harmonics)), tahb(std::move(tahb)), selectharms(std::move(selectharms)), startup_periods(std::move(startup_periods)), print_parameters(std::move(print_parameters)), nonlin_options(std::move(nonlin_options)), linsol_options(std::move(linsol_options)) {}

std::optional<HbSimulationParameters> HbSimulationParameters::from_xyce_directives(const std::vector<std::string>& directives) {
    // init defaults
    std::vector<std::string> frequencies;
    std::vector<int> harmonics;
    std::optional<int> tahb;
    std::optional<std::string> selectharms;
    std::optional<int> startup_periods;
    std::optional<PrintParameters> print_parameters;
    std::map<std::string, std::string> nonlin_options;
    std::map<std::string, std::string> linsol_options;

    // flag indicating whether a valid directive was found
    bool found = false;

    // parse directives
    for (const auto& directive : directives) {
        // tokenize the directive
        const auto tokens = tokenize(directive);

        // skip empty directives
        if (tokens.empty()) {
            continue;
        }

        const std::string cmd = to_upper(tokens[0]);

        // parse print directives and retain hb-specific output config
        if (cmd == ".PRINT") {
            // parse the print statement from the directive
            const auto print_statement = PrintParameters::from_xyce_statement(directive);
            // retain hb print parameters when found
            if (print_statement) {
                const std::string print_type_upper = to_upper(print_statement->print_type);
                if (print_type_upper == "HB" || print_type_upper == "HB_FD" || print_type_upper == "HB_TD") {
                    // store the parsed print parameters
                    print_parameters = *print_statement;
                }
            }
            continue;
        }

        // handle .OPTIONS
        if (cmd == ".OPTIONS" && tokens.size() > 1) {
            // get package
            const std::string pkg = to_upper(tokens[1]);

            // handle HBINT
            if (pkg == "HBINT") {
                // parse options
                for (size_t i = 2; i < tokens.size(); ++i) {
                    const auto& token = tokens[i];
                    // check for equals
                    const auto eq_pos = token.find('=');
                    if (eq_pos != std::string::npos) {
                        // split key and value
                        const auto key = to_upper(token.substr(0, eq_pos));
                        const auto val = token.substr(eq_pos + 1);

                        // handle NUMFREQ
                        if (key == "NUMFREQ") {
                            // split comma-separated values
                            std::string current_val;
                            for (size_t j = 0; j < val.length(); ++j) {
                                const char c = val[j];
                                if (c == ',') {
                                    // parse integer
                                    try {
                                        harmonics.push_back(std::stoi(current_val));
                                    }
                                    catch (...) {
                                        // ignore invalid NUMFREQ
                                    }
                                    current_val.clear();
                                }
                                else {
                                    current_val += c;
                                }
                            }
                            // parse last value
                            if (!current_val.empty()) {
                                try {
                                    harmonics.push_back(std::stoi(current_val));
                                }
                                catch (...) {
                                    // ignore
                                }
                            }
                        }
                        // handle TAHB
                        else if (key == "TAHB") {
                            // parse integer
                            try {
                                tahb = std::stoi(std::string(val));
                            }
                            catch (...) {
                                // ignore
                            }
                        }
                        // handle SELECTHARMS
                        else if (key == "SELECTHARMS") {
                            // store selectharms (lowercase)
                            selectharms = val;
                        }
                        // handle STARTUPPERIODS
                        else if (key == "STARTUPPERIODS") {
                            // parse integer
                            try {
                                startup_periods = std::stoi(std::string(val));
                            }
                            catch (...) {
                                // ignore
                            }
                        }
                    }
                }
                continue;
            }

            // handle NONLIN-HB
            if (pkg == "NONLIN-HB") {
                // parse options
                for (size_t i = 2; i < tokens.size(); ++i) {
                    const auto& token = tokens[i];
                    // check for equals
                    const auto eq_pos = token.find('=');
                    if (eq_pos != std::string::npos) {
                        // split key and value
                        const auto key = token.substr(0, eq_pos);
                        const auto val = token.substr(eq_pos + 1);
                        // store (uppercase key)
                        nonlin_options[to_upper(key)] = val;
                    }
                }
                continue;
            }

            // handle LINSOL-HB
            if (pkg == "LINSOL-HB") {
                // parse options
                for (size_t i = 2; i < tokens.size(); ++i) {
                    const auto& token = tokens[i];
                    // check for equals
                    const auto eq_pos = token.find('=');
                    if (eq_pos != std::string::npos) {
                        // split key and value
                        const auto key = token.substr(0, eq_pos);
                        const auto val = token.substr(eq_pos + 1);
                        // store (uppercase key)
                        linsol_options[to_upper(key)] = val;
                    }
                }
                continue;
            }
        }

        // skip non-HB directives
        if (cmd != ".HB") {
            continue;
        }

        // flag indicating a valid HB directive was found
        found = true;

        // collect all fundamental frequencies from remaining tokens
        for (size_t i = 1; i < tokens.size(); ++i) {
            frequencies.push_back(std::string(tokens[i]));
        }
    }

    // return instance if a valid directive was found
    if (!found) {
        return std::nullopt;
    }

    return HbSimulationParameters(frequencies, harmonics, tahb, selectharms, startup_periods, print_parameters, nonlin_options, linsol_options);
}

std::vector<std::string> HbSimulationParameters::to_xyce_directives(const NetlistTopology& topology) const {
    // init output directive list
    std::vector<std::string> directives;
    // topology reserved for future wildcard expansion; pass-through for now
    (void)topology;
    // build the hb directive with space-separated fundamental frequencies
    std::string hb_directive = ".HB";
    for (const auto& freq : frequencies) {
        hb_directive += " " + freq;
    }
    directives.push_back(hb_directive);

    // build hbint options
    std::vector<std::string> hbint_options;

    // append NUMFREQ when provided
    if (!harmonics.empty()) {
        // format as comma-separated integers
        std::string numfreq_str;
        for (size_t i = 0; i < harmonics.size(); ++i) {
            if (i > 0) {
                numfreq_str += ",";
            }
            numfreq_str += std::to_string(harmonics[i]);
        }
        hbint_options.push_back("NUMFREQ=" + numfreq_str);
    }

    // append TAHB when provided
    if (tahb.has_value()) {
        hbint_options.push_back("TAHB=" + std::to_string(*tahb));
    }

    // append SELECTHARMS when provided
    if (selectharms.has_value()) {
        hbint_options.push_back("SELECTHARMS=" + *selectharms);
    }

    // append STARTUPPERIODS when provided
    if (startup_periods.has_value()) {
        hbint_options.push_back("STARTUPPERIODS=" + std::to_string(*startup_periods));
    }

    // check if we have hbint options
    if (!hbint_options.empty()) {
        // append .OPTIONS HBINT line
        std::string hbint_directive = ".OPTIONS HBINT";
        for (const auto& opt : hbint_options) {
            hbint_directive += " " + opt;
        }
        directives.push_back(hbint_directive);
    }

    // append nonlin-hb options
    if (!nonlin_options.empty()) {
        std::string opts;
        for (const auto& [k, v] : nonlin_options) {
            if (!opts.empty()) {
                opts += " ";
            }
            opts += k + "=" + v;
        }
        directives.push_back(".OPTIONS NONLIN-HB " + opts);
    }

    // append linsol-hb options
    if (!linsol_options.empty()) {
        std::string opts;
        for (const auto& [k, v] : linsol_options) {
            if (!opts.empty()) {
                opts += " ";
            }
            opts += k + "=" + v;
        }
        directives.push_back(".OPTIONS LINSOL-HB " + opts);
    }

    // append hb print directive with topology-aware wildcard expansion
    if (print_parameters) {
        const std::string print_type_upper = to_upper(print_parameters->print_type);
        if (print_type_upper == "HB" || print_type_upper == "HB_FD" || print_type_upper == "HB_TD") {
            directives.push_back(print_parameters->to_xyce_statement());
        }
    }

    // return the full directive list
    return directives;
}

bool HbSimulationParameters::operator==(const HbSimulationParameters& other) const {
    // compare all fields for equality
    return frequencies == other.frequencies && harmonics == other.harmonics && tahb == other.tahb && selectharms == other.selectharms && startup_periods == other.startup_periods && print_parameters == other.print_parameters && nonlin_options == other.nonlin_options && linsol_options == other.linsol_options;
}
